docs/RFCS/20221124_split_tls.md

```diff
@@ -0,0 +1,295 @@
+copy of the file named `node.crt` to `server.crt` in the target
+certificate directory.
+
+A preliminary implementation is available here: https://github.com/cockroachdb/cockroach/pull/92126
+
+Topics that are out of scope for this RFC:
+
+- TLS client certs for authentication of SQL clients. These remain unchanged.
+- The "tenant client" cert, as used by standalone SQL servers when
+  connecting to each other or to the KV layer in CC serverless
+  deployments. This also will remain unchanged.
+- Refactoring the "certificate manager" code inside CockroachDB to not
+  assume specific paths / filenames for the various certs. While this
+  development would be desirable, we consider it to be orthogonal to
+  the current RFC.
+- Troubleshooting of TLS misconfiguration. This will be covered by a
+  separate RFC.
+
+# Motivation
+
+CockroachDB's TLS components support two main use cases to provision certificates:
+- a "simplistic" UX, which is the one we provide to low-knowledge
+  users, using CockroachDB's bundled 'cockroach cert' command to
+  produce certs, using self-signed CAs;
+- a more complete UX, where we acknowledge that good production
+  security relies on a deeper integration between CockroachDB and the
+  surrounding security infrastructure, here we allow users to
+  configure their own TLS certificates for use with CockroachDB.
+
+In this RFC we will consider the first case to be "convenience" and
+that the second case should be the primary design objective – i.e. we
+should make those database operators successful first. This case also
+includes our CockroachCloud team.
+
+A key characteristic of these customers is that they have their own
+PKI (Private Key Infrastructure), with their own rules about how
+certificates can be constructed. In other words, it's not reasonable
+for CockroachDB to place many restrictions on how certificates /
+certificate fields (especially Common Name, Subject Alternative Names,
+but not limited to these fields) can be populated before CockroachDB
+can use them.
+
+Within this context, our users face a couple of major issues:
+
+- The hostname/IP address for a node, as viewed from other nodes
+  (--listen-addr/–advertise-addr) may be different from that as seen
+  from SQL clients / RFC clients / web browsers for the UI. In
+  particular, the latter may not be determined before a crdb node has
+  finished initializing, e.g. when using automation tools to provision
+  networking infrastructure.
+
+- Certain IT orgs require a chain of human approval to approve new
+  service certs as visible across teams/departments. For example, when
+  deploying CockroachDB a manual approval may be needed when issuing
+  the TLS cert for the SQL service, because SQL apps are run by
+  different teams than the DB operators. In this type of organization,
+  it's very cumbersome when the same cert is also used by node-node
+  connections: every time a new node is added to the cluster (and a
```

Comment:
But wouldn't the new hostname for new node be required to be updated in the server cert for SQL apps anyways (assuming the separation)? And the kind of organizations we are referring to, both the clusters and apps would likely be within peered or same private networks - so what would be the issue in using wildcards for both node-to-node and server-to-SQL/HTTP/RPC?

I totally understand the motivation for separating the certs, but I'm not too sure about the reason here unless we've heard this from a customer directly.

Reply:
The SQL service is behind a LB and this is the hostname we can put in server.crt.

Reply:
> But wouldn't the new hostname for new node be required to be updated in the server cert for SQL apps anyways (assuming the separation)?

The hostname wouldn't even be necessary in the server certificates, since they just have to match the externally available DNS information. Having the same certificate for both the cluster-level hostname and the externally accessible domain name exposes internal cluster information to external clients, which can be undesirable depending on the threats the infrastructure is concerned with.

> And the kind of organizations we are referring to, both the clusters and apps would likely be within peered or same private networks - so what would be the issue in using wildcards for both node-to-node and server-to-SQL/HTTP/RPC?

As @knz said, this opens the door to using TCP load balancers, rather than having to have a proxy service or use peering to manage their network. For example, if an organization wants to be able to access their database cluster across cloud infrastructures, they would be able to connect to their cluster over TCP without having to migrate their database or infrastructure. I'd also argue it's not reasonable to assume that clusters and apps will be in the same private network or using peering, given that security conscious organizations are moving towards the zero trust model vs the network perimeter model